File: tools/binance/downloadBinancePriceHistory.js
import { writeFileSync, existsSync } from 'fs';
import { createWriteStream } from 'fs';
import https from 'https';

/**
 * Tool: Download Binance Price History
 * Downloads historical price data from Binance API
 */
export default {
  name: 'downloadBinancePriceHistory',
  description: 'Download historical cryptocurrency price data from Binance in CSV format',
  parameters: { 
    symbol: 'string',  // e.g., 'BTCUSDT'
    interval: 'string', // e.g., '1h', '1d', '15m'
    startTime: 'string|null', // Optional: ISO date or timestamp
    endTime: 'string|null',   // Optional: ISO date or timestamp
    limit: 'number',   // Optional: max 1000 per request
    outputFile: 'string' // Output CSV filename
  },
  
  async execute(params) {
    const { 
      symbol, 
      interval, 
      startTime = null, 
      endTime = null, 
      limit = 1000,
      outputFile 
    } = params;
    
    try {
      // Validate parameters
      const validIntervals = ['1m', '3m', '5m', '15m', '30m', '1h', '2h', '4h', '6h', '8h', '12h', '1d', '3d', '1w', '1M'];
      if (!validIntervals.includes(interval)) {
        throw new Error(`Invalid interval. Must be one of: ${validIntervals.join(', ')}`);
      }
      
      if (limit > 1000) {
        throw new Error('Limit cannot exceed 1000 per request');
      }

      // Build API URL
      let url = `https://api.binance.com/api/v3/klines?symbol=${symbol.toUpperCase()}&interval=${interval}&limit=${limit}`;
      
      if (startTime) {
        const startMs = this._parseTime(startTime);
        url += `&startTime=${startMs}`;
      }
      
      if (endTime) {
        const endMs = this._parseTime(endTime);
        url += `&endTime=${endMs}`;
      }

      console.log(`Downloading ${symbol} data from Binance (${interval} interval)...`);
      
      // Fetch data from Binance API
      const data = await this._fetchBinanceData(url);
      
      if (!data || data.length === 0) {
        throw new Error('No data received from Binance API');
      }

      // Convert to CSV format
      const csv = this._convertToCSV(data);
      
      // Write to file
      writeFileSync(outputFile, csv, 'utf-8');
      
      console.log(`✓ Downloaded ${data.length} candles to ${outputFile}`);
      
      return {
        success: true,
        symbol: symbol,
        interval: interval,
        records: data.length,
        filename: outputFile,
        startDate: new Date(data[0][0]).toISOString(),
        endDate: new Date(data[data.length - 1][0]).toISOString(),
        message: `Downloaded ${data.length} ${interval} candles for ${symbol}`
      };
      
    } catch (error) {
      return {
        success: false,
        error: error.message,
        message: `Failed to download Binance data: ${error.message}`
      };
    }
  },

  /**
   * Parse time string to milliseconds timestamp
   */
  _parseTime(timeStr) {
    // If already a number (timestamp), return it
    if (!isNaN(timeStr)) {
      return parseInt(timeStr);
    }
    
    // Parse ISO date string
    const date = new Date(timeStr);
    if (isNaN(date.getTime())) {
      throw new Error(`Invalid time format: ${timeStr}`);
    }
    
    return date.getTime();
  },

  /**
   * Fetch data from Binance API
   */
  async _fetchBinanceData(url) {
    return new Promise((resolve, reject) => {
      https.get(url, (res) => {
        let data = '';
        
        res.on('data', (chunk) => {
          data += chunk;
        });
        
        res.on('end', () => {
          try {
            const json = JSON.parse(data);
            
            // Check for API errors
            if (json.code && json.msg) {
              reject(new Error(`Binance API error: ${json.msg}`));
              return;
            }
            
            resolve(json);
          } catch (error) {
            reject(new Error(`Failed to parse Binance response: ${error.message}`));
          }
        });
      }).on('error', (error) => {
        reject(new Error(`HTTP request failed: ${error.message}`));
      });
    });
  },

  /**
   * Convert Binance klines data to CSV format
   * Binance klines format:
   * [
   *   [
   *     1499040000000,      // 0: Open time
   *     "0.01634000",       // 1: Open
   *     "0.80000000",       // 2: High
   *     "0.01575800",       // 3: Low
   *     "0.01577100",       // 4: Close
   *     "148976.11427815",  // 5: Volume
   *     1499644799999,      // 6: Close time
   *     "2434.19055334",    // 7: Quote asset volume
   *     308,                // 8: Number of trades
   *     "1756.87402397",    // 9: Taker buy base asset volume
   *     "28.46694368",      // 10: Taker buy quote asset volume
   *     "17928899.62484339" // 11: Ignore
   *   ]
   * ]
   */
  _convertToCSV(data) {
    // CSV header matching Binance format
    const header = 'timestamp,open,high,low,close,volume,close_time,quote_volume,trades,taker_buy_volume,taker_buy_quote_volume\n';
    
    const rows = data.map(candle => {
      return [
        candle[0],  // timestamp (open time)
        candle[1],  // open
        candle[2],  // high
        candle[3],  // low
        candle[4],  // close
        candle[5],  // volume
        candle[6],  // close time
        candle[7],  // quote volume
        candle[8],  // trades
        candle[9],  // taker buy volume
        candle[10]  // taker buy quote volume
      ].join(',');
    }).join('\n');
    
    return header + rows + '\n';
  }
};

File: obscurex/tools/binance/download_binance_price_history.py
"""
Tool: Download Binance Price History
Downloads historical price data from Binance API
"""

import requests
from datetime import datetime
from typing import Dict, Any, Optional, List


def _parse_time(time_str: str) -> int:
    """Parse time string to millisecond timestamp."""
    try:
        # If it's already a number, return it
        return int(time_str)
    except ValueError:
        # Try to parse as ISO date
        dt = datetime.fromisoformat(time_str.replace('Z', '+00:00'))
        return int(dt.timestamp() * 1000)


def _convert_to_csv(data: List) -> str:
    """Convert Binance kline data to CSV format."""
    # CSV header
    csv = "timestamp,open,high,low,close,volume,close_time,quote_volume,trades,taker_buy_volume,taker_buy_quote_volume\n"
    
    for candle in data:
        # Binance klines format: 
        # [open_time, open, high, low, close, volume, close_time, quote_asset_volume, 
        #  number_of_trades, taker_buy_base_volume, taker_buy_quote_volume, ignore]
        csv += f"{candle[0]},{candle[1]},{candle[2]},{candle[3]},{candle[4]},{candle[5]},"
        csv += f"{candle[6]},{candle[7]},{candle[8]},{candle[9]},{candle[10]}\n"
    
    return csv


async def execute(params: Dict[str, Any], context: Dict[str, Any] = None) -> Dict[str, Any]:
    """Download historical price data from Binance."""
    symbol = params.get('symbol')
    interval = params.get('interval')
    start_time = params.get('startTime')
    end_time = params.get('endTime')
    limit = params.get('limit', 1000)
    output_file = params.get('outputFile')
    
    try:
        # Validate parameters
        valid_intervals = ['1m', '3m', '5m', '15m', '30m', '1h', '2h', '4h', '6h', '8h', '12h', '1d', '3d', '1w', '1M']
        if interval not in valid_intervals:
            raise Exception(f"Invalid interval. Must be one of: {', '.join(valid_intervals)}")
        
        if limit > 1000:
            raise Exception('Limit cannot exceed 1000 per request')
        
        # Build API URL
        url = f"https://api.binance.com/api/v3/klines?symbol={symbol.upper()}&interval={interval}&limit={limit}"
        
        if start_time:
            start_ms = _parse_time(start_time)
            url += f"&startTime={start_ms}"
        
        if end_time:
            end_ms = _parse_time(end_time)
            url += f"&endTime={end_ms}"
        
        print(f"Downloading {symbol} data from Binance ({interval} interval)...")
        
        # Fetch data from Binance API
        response = requests.get(url, timeout=30)
        response.raise_for_status()
        data = response.json()
        
        if not data or len(data) == 0:
            raise Exception('No data received from Binance API')
        
        # Convert to CSV format
        csv = _convert_to_csv(data)
        
        # Write to file
        with open(output_file, 'w', encoding='utf-8') as f:
            f.write(csv)
        
        print(f"✓ Downloaded {len(data)} candles to {output_file}")
        
        return {
            'success': True,
            'symbol': symbol,
            'interval': interval,
            'records': len(data),
            'filename': output_file,
            'startDate': datetime.fromtimestamp(data[0][0] / 1000).isoformat(),
            'endDate': datetime.fromtimestamp(data[-1][0] / 1000).isoformat(),
            'message': f"Downloaded {len(data)} {interval} candles for {symbol}"
        }
    except requests.RequestException as e:
        raise Exception(f"Failed to download Binance data: {str(e)}")
    except Exception as e:
        raise Exception(f"Error downloading Binance data: {str(e)}")


# Tool definition
tool = {
    'name': 'downloadBinancePriceHistory',
    'description': 'Download historical cryptocurrency price data from Binance in CSV format',
    'parameters': {
        'symbol': 'string',
        'interval': 'string',
        'startTime': 'string|null',
        'endTime': 'string|null',
        'limit': 'number',
        'outputFile': 'string'
    },
    'execute': execute
}
